exploration/quoted-literals.md

```diff
@@ -0,0 +1,403 @@
+- [r2 GOOD] Embedding messages in certain container formats requires escaping the literal delimiters.
+  If the container format does not itself support dual quoting,
+  the embedded message's quotes may be adjusted to avoid their escaping.
+- [r3 GOOD] Quotation marks are universally recognized as string delimiters.
+- [r4 FAIR] Quotation marks cannot be paired by parsers nor IDEs,
+  but many text editors provide features to make working with and around quotes easier.
+
+### [a3] Use round or angle brackets
+
+- Round parentheses are very uncommon as string delimiters [r2 GOOD],
+  and thus may be surprising,
+  especially given the well-established meaning in prose [r4 POOR].
+  That said, there's prior art in using them for [delimiting strings in PostScript](https://en.wikipedia.org/wiki/PostScript#%22Hello_world%22).
+  Furthermore, they are relatively common in text, where they'd require escaping [r1 POOR].
+- Angle brackets require escaping in XML-based storage formats [r2 FAIR].
+- All brackets can be easily paired by parsers and IDEs [r5 GOOD].
+
+### [a4] Change escape introducer
+
+Changing the escape sequence introducer from backslash [c1] to another character
+could help partially mitigate the burden of first escaping literal delimiters
+and then escaping the escapes themselves [r1].
+However, it wouldn't address other requirements and use-cases.
+
+### [a5] Double delimiters to escape them
+
+This is the approach taken by ICU MessageFormat 1.0 for quotes.
+It allows literals to contain quotes [r1 GOOD]
+at the expense of doubling the amount of escaping required when embedding messages in code [r2 POOR].
+
+### [a6] Accept either `|` or quotes
+
+Allow any of the following as literal delimiters:
+
+- the vertical line character `|`
+- single quotes `'`
+- double quotes `"`
+
+This approach supports multiple different quoting styles to be used for literals.
+This flexibility allows for using a familiar and common style such as `'single'` or `"double"` quotes,
+while also allowing for `|pipes|` when the message's contents or embedding would otherwise require additional escaping.
+
+```abnf
+literal       = quoted / unquoted
+quoted        = "|" *(quoted-char / "'" / DQUOTE / quoted-escape) "|"
+              / "'" *(quoted-char / DQUOTE / "|" / quoted-escape) "'"
+              / DQUOTE *(quoted-char / "'" / "|" / quoted-escape) DQUOTE
+quoted-char   = %x0-21         ; omit "
+              / %x23-26        ; omit '
+              / %x28-5B        ; omit \
+              / %x5D-7B        ; omit |
+              / %x7D-D7FF      ; omit surrogates
+              / %xE000-10FFFF
+quoted-escape = backslash ( backslash / "|" / "'" / DQUOTE )
+```
+
+- [r1 GOOD] Writing any two of `|`, `"` and `'` in literals doesn't require escaping them via `\`.
+  This means no extra `\` that need escaping.
+  Message don't have to be modified otherwise before embedding them,
+  unless they happen to contain conflicting quote delimiters.
```

Comment:
This is actually a con "in a trenchcoat" :-)

Messages MUST be modified if they "happen to contain conflicting quote delimiters", e.g. those that conflict with the syntax.

My big concern here is: how to do tools up and down the translation stack decide what quotes to use?

There should be a "con" somewhere for having multiple ways of doing the same thing.

Reply:
That's `r6`, and it's included in the table.

Reply:
Oops, sorry, typo above: I meant `r5`, the "one way" requirement.